### ISSUE TITLE
Clustering visualization and metrics

### ISSUE THREAD

**Issue by ShailenSmith** [LukasStruppek/Plug-and-Play-Attacks#15]
Hi,

My team and I are interested in replicating the t-SNE and violin plots (Figure 5) in the NegLS paper. Could you please clarify the following:

- For the t-SNE plots, what is meant by 'FaceNet classes' ? What dataset does this refer to, and how were the 100 classes chosen?

- For the t-SNE plots, what visualization hyperparameters were chosen?

- Out of curiosity, did you happen to try using UMAP for the same purpose?

- For the violin plots, you mention that the L2 feature distances are 'maximum-scaled'. Are the L2 distances for each of the three models divided by the respective max pairwise L2 distance for the model, or are all three sets of L2 distances divided by the max pairwise L2 distance across all the models?

- Is max-normalization necessary here, i.e., are the metrics for the three different models comparable without normalizing their distances? (My first thought is yes, because the feature embedding space is the same for each of the models, but I'm not very familiar with using clustering metrics in this way, so I could be wrong.)

- For the violin plots, which train set was used?

Thanks for your excellent work, and for making your code publicly available! Looking forward to hearing from y'all.

**Comment by LukasStruppek** [comment]
Hi Shailen,

Thank you for your interest in our research. Regarding your questions:
- FaceNet classes: Oh, you spot a typo here. It is actually FaceScrub classes, of which we used the first 50 actors + 50 actresses (alphabetically ordered)
- t-SNE hyperparameters: n_iter=1000, perplexity=30, random_state=1, n_components=2 for sklearn-manifold.TSNE. I hope those are the correct parameters, it may be the case that I changed that before committing my code.
- No, we only tested tsne. But UMAP also sounds reasonable.
- Exactly, we divided the values by their maximum to limit the scaling at 1 which is necessary to get all plots in the same range. Otherwise, models trained with pos/neg. LS have much higher absolute distances compared to the model trained with hard labels.
- for tSNE, no scaling is required (we also didn't apply any there). But to put the violin plots on the same scale, we needed this type of scaling. Instead, you could also use absolute distances with different axes for the three settings (pos LS, neg. LS, hard labels)
- All experiemnts starting from 4.2 are based on the ResNet-152 trained on FaceScrub (models are provided in the Repo here). So both, the tSNE and the violin plots, also use these models, including the FaceScrub training data.

I hope this answers your questions. If there are any more, feel free to reach out to me :)

Best,
Lukas

**Comment by ShailenSmith** [comment]
Hi Lukas,

Thanks for your response, that's very helpful.

In my own testing, I see that while negLS embeddings have absolute distances that are much higher than the hard-label model, the posLS embeddings are actually much lower. I am using ResNet-152 on all FaceScrub classes.

Is there anything to be said for the fact that the absolute distances for pos/neg LS embeddings are very different from those of the hard-label model? Do you think there's any merit to a claim like 'smaller absolute distances between inter-class embeddings makes model inversion easier'? From my current understanding, that could be an alternate explanation for why label smoothing has the effect that it does on model inversion.

**Comment by LukasStruppek** [comment]
Hmm, that is an interesting point you raise. I am not 100% sure that smaller distances automatically increase the privacy leakage but it might be worth looking into it. Maybe also have a look at https://arxiv.org/abs/1906.02629, which explored the behavior of label smoothing in general. To test your hypothesis, you could also try training models with another regularization to force the model to have a compact embedding space and then check if the privacy leakage indeed increases.

If you are interested in having a chat about that, I am always happy to meet.

**Comment by ShailenSmith** [comment]
Thanks for the advice and the paper link, looks interesting, and I appreciate the offer to chat! I'll let you know if anything worth discussing comes out of my investigations.
